GPU/D3D11/FramebufferManagerD3D11.cpp

```diff
@@ -0,0 +1,1012 @@
+	}
+}
+
+void FramebufferManagerD3D11::ReformatFramebufferFrom(VirtualFramebuffer *vfb, GEBufferFormat old) {
+	if (!useBufferedRendering_ || !vfb->fbo) {
+		return;
+	}
+
+	draw_->BindFramebufferAsRenderTarget(vfb->fbo);
+
+	// Technically, we should at this point re-interpret the bytes of the old format to the new.
+	// That might get tricky, and could cause unnecessary slowness in some games.
+	// For now, we just clear alpha/stencil from 565, which fixes shadow issues in Kingdom Hearts.
+	// (it uses 565 to write zeros to the buffer, than 4444 to actually render the shadow.)
+	//
+	// The best way to do this may ultimately be to create a new FBO (combine with any resize?)
+	// and blit with a shader to that, then replace the FBO on vfb.  Stencil would still be complex
+	// to exactly reproduce in 4444 and 8888 formats.
+
+	if (old == GE_FORMAT_565) {
+		context_->OMSetDepthStencilState(stockD3D11.depthDisabledStencilWrite, 0xFF);
+		context_->OMSetBlendState(stockD3D11.blendStateDisabledWithColorMask[0], nullptr, 0xFFFFFFFF);
+		context_->RSSetState(stockD3D11.rasterStateNoCull);
+		context_->IASetInputLayout(quadInputLayout_);
+		context_->PSSetShader(quadPixelShader_, nullptr, 0);
+		context_->VSSetShader(quadVertexShader_, nullptr, 0);
+		context_->IASetVertexBuffers(0, 1, &fsQuadBuffer_, &quadStride_, &quadOffset_);
+		shaderManager_->DirtyLastShader();
+		D3D11_VIEWPORT vp{ 0.0f, 0.0f, (float)vfb->renderWidth, (float)vfb->renderHeight, 0.0f, 1.0f };
+		context_->RSSetViewports(1, &vp);
+		context_->IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP);
+		context_->Draw(2, 0);
+	}
+
+	RebindFramebuffer();
+}
+
+static void CopyPixelDepthOnly(u32 *dstp, const u32 *srcp, size_t c) {
+	size_t x = 0;
+
+#ifdef _M_SSE
+	size_t sseSize = (c / 4) * 4;
+	const __m128i srcMask = _mm_set1_epi32(0x00FFFFFF);
+	const __m128i dstMask = _mm_set1_epi32(0xFF000000);
+	__m128i *dst = (__m128i *)dstp;
+	const __m128i *src = (const __m128i *)srcp;
+
+	for (; x < sseSize; x += 4) {
+		const __m128i bits24 = _mm_and_si128(_mm_load_si128(src), srcMask);
+		const __m128i bits8 = _mm_and_si128(_mm_load_si128(dst), dstMask);
+		_mm_store_si128(dst, _mm_or_si128(bits24, bits8));
+		dst++;
+		src++;
+	}
+#endif
+
+	// Copy the remaining pixels that didn't fit in SSE.
+	for (; x < c; ++x) {
+		memcpy(dstp + x, srcp + x, 3);
+	}
+}
+
+void FramebufferManagerD3D11::BlitFramebufferDepth(VirtualFramebuffer *src, VirtualFramebuffer *dst) {
+	if (g_Config.bDisableSlowFramebufEffects) {
+		return;
+	}
+	bool matchingDepthBuffer = src->z_address == dst->z_address && src->z_stride != 0 && dst->z_stride != 0;
+	bool matchingSize = src->width == dst->width && src->height == dst->height;
+	if (matchingDepthBuffer && matchingSize) {
+		// Doesn't work.  Use a shader maybe?
+		draw_->BindBackbufferAsRenderTarget();
+		draw_->CopyFramebufferImage(src->fbo, 0, 0, 0, 0, dst->fbo, 0, 0, 0, 0, src->renderWidth, src->renderHeight, 1, Draw::FB_DEPTH_BIT);
```

Comment:
Doesn't this also copy stencil, which isn't necessarily desirable (since the PSP keeps stencil with the color buffer, which hasn't been switched)?  This would also apply to using StretchRect on D3D9, right?

-[Unknown]